db/impl/etcd/client.go

```diff
@@ -0,0 +1,291 @@
+			return errors.EtcdToErrored(err)
+		}
+	}
+
+	path, err := obj.Path()
+	if err != nil {
+		return err
+	}
+
+	if _, err := c.client.Delete(context.Background(), c.qualified(path), nil); err != nil {
+		return errors.EtcdToErrored(err)
+	}
+
+	if obj.Hooks().PostDelete != nil {
+		if err := obj.Hooks().PostDelete(c, obj); err != nil {
+			return errors.EtcdToErrored(err)
+		}
+	}
+
+	return nil
+}
+
+// Prefix returns a copy of the string used to make the database prefix.
+func (c *Client) Prefix() string {
+	return c.prefix
+}
+
+// Watch watches a given object for changes.
+func (c *Client) Watch(obj db.Entity) (chan db.Entity, chan error) {
+	path, err := obj.Path()
+	if err != nil {
+		errChan := make(chan error, 1)
+		errChan <- err
+		return make(chan db.Entity), errChan
+	}
+
+	return c.watchPath(obj, path, false)
+}
+
+// WatchStop stops a watch for a given object.
+func (c *Client) WatchStop(obj db.Entity) error {
+	path, err := obj.Path()
+	if err != nil {
+		return err
+	}
+
+	return c.watchStopPath(path)
+}
+
+// WatchAll watches all items under the given entity's
+func (c *Client) WatchAll(obj db.Entity) (chan db.Entity, chan error) {
+	return c.watchPath(obj, obj.Prefix(), true)
+}
+
+// WatchAllStop stops
+func (c *Client) WatchAllStop(obj db.Entity) error {
+	return c.watchStopPath(obj.Prefix())
+}
+
+// WatchPath watches an object for changes. Returns two channels: one for entity updates and one for errors.
+// You can stop watches with WatchStop with the same path.
+// Only one watch for a given path may be active at a time.
+func (c *Client) watchPath(obj db.Entity, path string, recursive bool) (chan db.Entity, chan error) {
+	c.watcherMutex.Lock()
+	defer c.watcherMutex.Unlock()
+
+	stopChan := make(chan struct{}, 1)
+	retChan := make(chan db.Entity)
+	errChan := make(chan error, 1)
+
+	go func() {
+		watcher := c.client.Watcher(c.qualified(path), &client.WatcherOptions{Recursive: recursive})
+
+		ctx, cancel := context.WithCancel(context.Background())
+		go func() {
+			<-stopChan
+			cancel()
+		}()
+
+		for {
+			resp, err := watcher.Next(ctx)
+			if err != nil {
+				if err == context.Canceled {
+					logrus.Debugf("watch for %q canceled", path)
+					return
+				}
+
+				errChan <- err
+
+				time.Sleep(time.Second)
+				continue
+			}
+
+			for _, entity := range c.traverse(resp.Node, obj) {
+				retChan <- entity
+			}
+		}
+	}()
+
+	_, ok := c.watchers[path]
+	if ok {
+		close(c.watchers[path])
+	}
+	c.watchers[path] = stopChan
+
+	return retChan, errChan
+}
+
+// WatchStopPath stops a watch given a path to stop the watch on.
+func (c *Client) watchStopPath(path string) error {
+	c.watcherMutex.Lock()
+	defer c.watcherMutex.Unlock()
+
+	stopChan, ok := c.watchers[path]
+	if !ok {
+		return errors.InvalidDBPath.Combine(errored.New("missing key during watch"))
+	}
+
+	close(stopChan)
+	delete(c.watchers, path)
+
+	return nil
+}
+
+func (c *Client) traverse(node *client.Node, obj db.Entity) []db.Entity {
```

Comment:
it traverses the keyspace and turns anything that looks like a passed entity into an entity to put into the list to return.

I'll put in a comment in v2.